### PR TITLE
[IMP] account{,_peppol}: Add Peppol information in mail footer

### PR DESCRIPTION
**Pull request by clbr-odoo**
Add information in email footer when relevant to let people know that Peppol will become mandatory in their country, and recommend Odoo as a compliant software in that regard.

task-4332306

Example of sent emails:
To a customer in a Peppol country, with Email + Peppol sending:
image

To a customer in a Peppol country with Email only (obviously the Belgian specific information only appear to invoices sent to belgian customers):
image

To a cutomer in a non Peppol country: No change
image

### CODE REVIEW

File: addons/account/models/company.py [hunk 1]
```diff
@@ -28,12 +28,19 @@
...
     'AD', 'AL', 'AT', 'BA', 'BE', 'BG', 'CH', 'CY', 'CZ', 'DE', 'DK', 'EE', 'ES', 'FI',
     'FR', 'GB', 'GR', 'HR', 'HU', 'IE', 'IS', 'IT', 'LI', 'LT', 'LU', 'LV', 'MC', 'ME',
     'MK', 'MT', 'NL', 'NO', 'PL', 'PT', 'RO', 'RS', 'SE', 'SI', 'SK', 'SM', 'TR', 'VA',
 ]
 
+PEPPOL_DEFAULT_COUNTRIES = [  # List of countries where Peppol should be used by default.
+    'AT', 'BE', 'CH', 'CY', 'CZ', 'DE', 'DK', 'EE', 'ES', 'FI',
+    'FR', 'GR', 'IE', 'IS', 'IT', 'LT', 'LU', 'LV', 'MT', 'NL',
+    'NO', 'PL', 'PT', 'RO', 'SE', 'SI',
+]
```
Comment (smetl) [inline addons/account/models/company.py:41]:
```suggestion
# List of countries where Peppol should be used by default.
PEPPOL_DEFAULT_COUNTRIES = [
    'AT', 'BE', 'CH', 'CY', 'CZ', 'DE', 'DK', 'EE', 'ES', 'FI',
    'FR', 'GR', 'IE', 'IS', 'IT', 'LT', 'LU', 'LV', 'MT', 'NL',
    'NO', 'PL', 'PT', 'RO', 'SE', 'SI',
]

# List of countries where Peppol is accessible.
PEPPOL_LIST = ['AD', 'AL',  'BA', 'BG', 'GB', 'HR', 'HU', 'LI', 'MC', 'ME', 'MK', 'RS', 'SK', 'SM', 'TR', 'VA'] + PEPPOL_DEFAULT_COUNTRIES
```

File: addons/account_peppol/models/account_move.py [hunk 2]
```diff
@@ -70,3 +71,18 @@ def _compute_peppol_move_state(self):
...
+    def _notify_by_email_prepare_rendering_context(self, message, msg_vals=False, model_description=False, force_email_company=False, force_email_lang=False):
+        render_context = super()._notify_by_email_prepare_rendering_context(
+            message, msg_vals, model_description=model_description,
+            force_email_company=force_email_company, force_email_lang=force_email_lang
+        )
+        if model_description == 'Invoice':
+            invoice = render_context['record']
+            is_peppol_country = invoice.commercial_partner_id.country_code in PEPPOL_DEFAULT_COUNTRIES
+            sent_by_peppol = invoice.peppol_move_state in ('processing', 'done')
+            render_context['peppol_ad'] = {
```
Comment (smetl) [inline addons/account_peppol/models/account_move.py:84]: `peppol_ad`... `ad`.... :thinking: ... `additional data` ? ... `anno domini` ? :confused:

File: addons/account_peppol/models/account_move.py [hunk 3]
```diff
@@ -70,3 +71,18 @@ def _compute_peppol_move_state(self):
...
+                'is_peppol_country': is_peppol_country,
+                'is_peppol_sent': sent_by_peppol,
+            }
```
Comment (smetl) [inline addons/account_peppol/models/account_move.py:87]:
```suggestion
            peppol_country = invoice.commercial_partner_id.country_code
            is_peppol_country = peppol_country in PEPPOL_DEFAULT_COUNTRIES
            render_context['peppol_ad'] = {
                'peppol_country': peppol_country,  # For the Belgium check in the view ;)
                'is_peppol_country': peppol_country in PEPPOL_DEFAULT_COUNTRIES,
                'is_peppol_sent': invoice.peppol_move_state in ('processing', 'done'),
            }
```

File: addons/account_peppol/models/account_move.py [hunk 4]
```diff
@@ -70,3 +71,18 @@ def _compute_peppol_move_state(self):
...
+
+    def _notify_by_email_prepare_rendering_context(self, message, msg_vals=False, model_description=False, force_email_company=False, force_email_lang=False):
+        render_context = super()._notify_by_email_prepare_rendering_context(
+            message, msg_vals, model_description=model_description,
+            force_email_company=force_email_company, force_email_lang=force_email_lang
+        )
+        if model_description == 'Invoice':
+            invoice = render_context['record']
+            invoice_country = invoice.commercial_partner_id.country_code
+            if invoice_country in PEPPOL_DEFAULT_COUNTRIES:
```
Comment (clbr-odoo) [inline addons/account_peppol/models/account_move.py:83]: @smetl I changed it even a bit more further in the direction of your suggestion.
Then I don't need to know if "is_peppol_country", the info will just not be there if its not.

File: addons/account_peppol/models/account_move.py [hunk 5]
```diff
@@ -70,3 +71,18 @@ def _compute_peppol_move_state(self):
                 move.peppol_move_state = False
             else:
                 move.peppol_move_state = move.peppol_move_state
+
+    def _notify_by_email_prepare_rendering_context(self, message, msg_vals=False, model_description=False, force_email_company=False, force_email_lang=False):
+        render_context = super()._notify_by_email_prepare_rendering_context(
+            message, msg_vals, model_description=model_description,
+            force_email_company=force_email_company, force_email_lang=force_email_lang
+        )
+        if model_description == 'Invoice':
```
Comment (tde-banana-odoo) [inline addons/account_peppol/models/account_move.py:80]: This is normally translated, so I suppose it is going to be broken, isn't it ?

Comment (tde-banana-odoo) [reply addons/account_peppol/models/account_move.py:80]: model description being "Invoice" is probably the same as considering self._name == "account.move", which is always the case as you override account_move model, isn't it ?

File: addons/account_peppol/models/account_move.py [hunk 6]
```diff
@@ -70,3 +71,18 @@ def _compute_peppol_move_state(self):
                 move.peppol_move_state = False
             else:
                 move.peppol_move_state = move.peppol_move_state
+
+    def _notify_by_email_prepare_rendering_context(self, message, msg_vals=False, model_description=False, force_email_company=False, force_email_lang=False):
```
Comment (tde-banana-odoo) [inline addons/account_peppol/models/account_move.py:75]: Could simplify the override (and ease changes when changing the base method)
```suggestion
    def _notify_by_email_prepare_rendering_context(self, *, model_description=False, **kwargs):
```

**Review by tde-banana-odoo — COMMENTED** [review]
Saw some light :wave:

**Comment by clbr-odoo** [comment]
@tde-banana-odoo thank you, very helpful. Is my `super()` call still correct with the `*` notation ? I'm not used to it.

File: addons/account_peppol/models/account_move.py (hunk 6 above)
Comment (clbr-odoo) [reply addons/account_peppol/models/account_move.py]: Heh, runbot doesn't agree with the change, will have a look a it.

File: addons/account_peppol/models/account_move.py [hunk 7]
```diff
@@ -70,3 +71,14 @@ def _compute_peppol_move_state(self):
...
                 move.peppol_move_state = move.peppol_move_state
+
+    def _notify_by_email_prepare_rendering_context(self, message, **kwargs):
+        render_context = super()._notify_by_email_prepare_rendering_context(message, **kwargs)
+        invoice = render_context['record']
+        invoice_country = invoice.commercial_partner_id.country_code
+        if invoice_country in PEPPOL_DEFAULT_COUNTRIES:
+            render_context['peppol_info'] = {
+                'peppol_country': invoice_country,
+                'is_peppol_sent':  invoice.peppol_move_state in ('processing', 'done'),
```
Comment (smetl) [inline addons/account_peppol/models/account_move.py:82]:
```suggestion
                'is_peppol_sent': invoice.peppol_move_state in ('processing', 'done'),
```

**Review by smetl — CHANGES_REQUESTED** [review]
How dare you! :grimacing:

**Review by smetl — APPROVED** [review]

**Comment by d-fence** [comment]
Upgrade exception [#409](https://runbot.odoo.com/web/#id=409&view_type=form&model=runbot.upgrade.exception) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

PRs adding fields/models should be merged at the beginning of the week and all the forward ports should be merged in the same week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
view:account_peppol.mail_notification_layout_with_responsible_signature_and_peppol
```
cc @KangOl @nseinlet @aj-fuentes

**Review by smetl — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**